src/binxml/assemble.rs

```diff
@@ -33,144 +34,283 @@ pub fn parse_tokens<W: Write, T: BinXmlOutput<W>>(
     Ok(())
 }
 
-pub fn create_record_model(tokens: Vec<BinXMLDeserializedTokens>) -> Vec<XmlModel> {
+pub fn create_record_model<'a>(tokens: Vec<Cow<'a, BinXMLDeserializedTokens<'a>>>) -> Vec<XmlModel<'a>> {
     let mut current_element: Option<XmlElementBuilder> = None;
-    let mut model: Vec<XmlModel> = vec![];
+    let mut model: Vec<XmlModel> = Vec::with_capacity(tokens.len());
 
     for token in tokens {
         match token {
-            BinXMLDeserializedTokens::FragmentHeader(_) => {}
-            BinXMLDeserializedTokens::TemplateInstance(_) => {
-                panic!("Call `expand_templates` before calling this function")
-            }
-            BinXMLDeserializedTokens::AttributeList => {}
-            BinXMLDeserializedTokens::Attribute(attr) => {
-                trace!("BinXMLDeserializedTokens::Attribute(attr) - {:?}", attr);
-                match current_element.take() {
-                    None => panic!("attribute - Bad parser state"),
-                    Some(builder) => {
-                        current_element = Some(builder.attribute_name(attr.name));
-                    }
-                };
-            }
-            BinXMLDeserializedTokens::OpenStartElement(elem) => {
-                trace!(
-                    "BinXMLDeserializedTokens::OpenStartElement(elem) - {:?}",
-                    elem.name
-                );
-                let builder = XmlElementBuilder::new();
-                current_element = Some(builder.name(elem.name));
-            }
-            BinXMLDeserializedTokens::CloseStartElement => {
-                trace!("BinXMLDeserializedTokens::CloseStartElement");
-                match current_element.take() {
-                    None => panic!("close start - Bad parser state"),
-                    Some(builder) => model.push(XmlModel::OpenElement(builder.finish())),
-                };
-            }
-            BinXMLDeserializedTokens::CloseEmptyElement => {
-                trace!("BinXMLDeserializedTokens::CloseEmptyElement");
-                match current_element.take() {
-                    None => panic!("close empty - Bad parser state"),
-                    Some(builder) => {
-                        model.push(XmlModel::OpenElement(builder.finish()));
+            Cow::Owned(owned_token) => {
```

Comment:
I’m not sure I get why everything is duplicated in these two branches. What does having a `Cow::Owned(RefValue)` means?

Perhaps we could reduce duplication here by some better pattern matching?